Lock screen using UIWindow and blurring

tssmith7 opened the pull request:
I have read that the Lock Screen has some issues when a modal view is on screen.  I was thinking of working on the Document Picker code and the lock screen issue came up as a potential serious problem with the implementation of it (see  #389).  So, I decided to try and come up with a version that works better than the existing code.

This is what happens now if you have the Export menu open and switch apps, return, and enter the pin:
![img_2293](https://user-images.githubusercontent.com/26643468/26951910-f2c4b222-4c57-11e7-9ccf-ea12fd994fbf.PNG)

This happens if you have the "Add Group/Entry" menu open and switch apps:
![img_2294](https://user-images.githubusercontent.com/26643468/26951931-09114482-4c58-11e7-9c38-705b7d542f09.PNG)

Other very strange things happen if you use the touchID.  Sometimes it dismisses the view that was on the screen before locking etc.

I am not sure but I think the problem with this is the use of the Navigation Controller.  I tried something different.  I created a new UIWindow that has a blur effect for the content below it.  Initially I tried it with the existing v1.6.3 code and it had trouble with the UIActionSheetAutoDismiss code.  It turns out that the auto-dismissing of those UIActionSheets never occurs (they are also depreciated now).  That was because init() was overridden and only initWithTitle() was ever called.  I fixed that issue and it worked great.  However, since the new master branch is partially converted to Swift, it doesn't call the UIActionSheetAutoDismiss class at all.  So it works good with the Swift code and I didn't include the UIActionSheetAutoDismiss class changes.

Also, the lock screen used to respond to the UIApplicationWillResignActiveNotification and UIApplicationDidBecomeActiveNotification.  These are fired off as the touchID screen comes up which made some of the logic more complicated.  I switched to using UIApplicationWillEnterForegroundNotification and UIApplicationDidEnterBackgroundNotification as suggested here [(Technical Q&A QA1838)](https://developer.apple.com/library/content/qa/qa1838/_index.html).

Let me know if this works for you.
![img_2295](https://user-images.githubusercontent.com/26643468/26952615-1f899b9e-4c5b-11e7-9835-73216c17d952.PNG)

jarush commented:
@tallerthenyou - check this out.

I like the look of it.  We'll have to do a lot of testing, the lock screen has notoriously been a tricky thing to get to work under all corner cases.  You mentioned that this doesn't work with UIActionSheet or the broken UIActionSheetAutoDismiss classes.  I honestly don't even recall if we use either of these in the new swift code, but if we do we'll need all the fixes in place before we would consider merging this in.

I'll try and do a code review of this when I get a chance, but I noticed a couple of minor things I'd ask you to look at when submitting pull requests.
1. Could you try and match the code format of your changes to the format in the file you're updating?  For the most part this hasn't been an issue, but I've noticed some discrepancies with white-space (particularly around if (...) formatting).
2. When submitting pull requests, could you remove code that you commented out during development?

Thanks for the pull request!

jarush commented on MiniKeePass/Lock Screen/KeypadView.m:
```diff
@@ -27,7 +27,7 @@ @implementation KeypadView
     self = [super initWithFrame:frame];
     if (self) {
-        self.backgroundColor = [UIColor whiteColor];
+//        self.backgroundColor = [UIColor whiteColor];
```
Delete the commented out code.

jarush commented on MiniKeePass/Lock Screen/LockScreenManager.m:
```diff
@@ -15,6 +15,12 @@
+// NOTE: This will NOT work with the old UIActionSheet depreciated class.
+//       Also, the UIActionSheetAutoDismiss class is broken and doesn't auto dismiss
+//       unless fixed.  If fixed then it works.
+
```
We should update the code to get rid of all the old UIActionSheet or UIActionSheetAutoDismiss classes.

tssmith7 replied:
I removed them as they were not used.  Should I remove this NOTE: as well?

jarush replied:
You could remove the note.  We shouldn't be using the old UIActionSheet class anymore.

jarush commented on MiniKeePass/Lock Screen/LockScreenManager.m:
```diff
@@ -26,13 +32,14 @@
 #import "PasswordUtils.h"
 
 @interface LockScreenManager () <PinViewControllerDelegate>
-@property (nonatomic, strong) LockViewController *lockViewController;
```
It looks like your implementation no longer uses the LockViewController class.  If so, that class should be deleted and the deletion included in the pull request.

jarush commented on MiniKeePass/Lock Screen/LockScreenManager.m:
```diff
@@ -90,6 +107,10 @@ - (BOOL)shouldCheckPin {
         return NO;
     }
+    
+    if( touchIDFailed ) {
```
Fix formatting

jarush commented on MiniKeePass/Lock Screen/LockScreenManager.m:
```diff
@@ -103,81 +124,44 @@ - (BOOL)shouldCheckPin {
+    // Perform Touch ID if enabled and not already failed.
     AppSettings *appSettings = [AppSettings sharedInstance];
-    if ([appSettings touchIdEnabled]) {
+    if ([appSettings touchIdEnabled] && !touchIDFailed ) {
```
Fix formatting

jarush commented on MiniKeePass/Lock Screen/PinViewController.m:
```diff
@@ -31,7 +31,7 @@ @implementation PinViewController
     self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
     if (self) {
-        self.view.backgroundColor = [UIColor whiteColor];
+//        self.view.backgroundColor = [UIColor whiteColor];
```
Delete commented out code.

jarush commented on MiniKeePass/Settings View/SettingsViewController.swift:
```diff
@@ -226,6 +226,7 @@ class SettingsViewController: UITableViewController, PinViewControllerDelegate {
             let pinViewController = PinViewController()
             pinViewController.titleLabel.text = NSLocalizedString("Set PIN", comment: "")
             pinViewController.delegate = self
+            pinViewController.view.backgroundColor = UIColor.white
```
Maybe add a short comment here describing why you're setting the background to white in the SettingsViewController.  It took me a little while to realize that it was because we wanted the background white for setting the PIN, but transparent so the UIWindow could blur it out in the lock screen.

tssmith7 commented:
The new Swift code does not use UIActionSheet or UIActionSheetAutoDismiss.  I removed the UIActionSheetAutoDismiss class.  I also removed LockViewController as it is not used anymore as you pointed out.

I changed the formatting and removed commented out code. 

The way this lock screen looks and behaves is a little different than the way the old one did.  The pin entry keypad is always shown and is behind the touchID dialog when touchID is enabled.  

If you have the pin disabled and "Close Database On Timeout" enabled then the pin entry keypad is shown briefly and fades away when the user brings the app into the foreground.  I don't understand why this is the case.  Is it a bug related to #490?

Let me know your thoughts.

jarush commented:
The changes look good, thank you!

The PIN shouldn't show up if it's not enabled, even if close database on timeout is enabled.  It sounds like there's a bug in there somewhere.

tssmith7 commented:
I fixed the pin screen showing up when not enabled.  Also fixed that the exitTime wasn't set if the pin was disabled.

jarush commented:
Yeah, the contents of MiniKeePass should always be hidden from the task switcher and if the database is going to be closed for any reason.  I believe that's why we had the the LockViewController, to hide the contents of the app if the PIN or Lock on Timeout was enabled.  If the PIN was enabled, it would show the PinViewController, but if it wasn't enabled it would remove the LockViewController after the app resumed.  This gave the app enough time to close the DB and hide any personal information before the LockViewController was remove.

Basically, we hide any personal information that could possibly be shown if for any reason the app wasn't going to resume and just show them the content anyways.  Make sense?

tssmith7 commented:
Changed it so it always blurs the task switcher window.  The pin keypad is only shown when the pin is enabled.  The blurred screen is removed such that if the database is closed nothing is discernible until after it has closed. I believe it works as desired now.

tallerthenyou commented:
This looks really great!
I'd originally shied away from using a UIWindow for the lock screen, but given all the trouble that the modal view has given us it's probably the right call.

I noticed that's the contents MiniKeePass aren't covered in the app switcher if MKP is the app you're switching from. I'm personally ok with the change because MKP was just full screen a split second before that, but do we think users will notice/care about the change?

Super trivial addition. Could you make the blurView or the lockWindow also alpha away with the pinViewController? I tested fading out the blurView alongside the pinViewController and I liked the look of it.

tssmith7 commented:
I didn't not a difference with both the pin view and the blur fading versus just the window fading so I changed it to just the window fading.  I (or you) can change it as you like.
